### ISSUE TITLE
VS Code «failed to load» even in cocalc-docker

### ISSUE THREAD

**Issue by belonesox** [sagemathinc/cocalc#7009]
I tried to refresh my cocalc installation and encounter problem:
* Pressing «VS Code» button results pause then «failed to load»
* code-server running without problem (empty `stderr`, good «listening port» in `stdout`)
* I reproduced it even in cocalc-docker
![image](https://github.com/sagemathinc/cocalc/assets/1609739/c61287ed-8379-4f45-805b-0583838e8b60)

I tried to debug it, looks on `packages/project` side all OK, 
* code-server runned successfully 

![image](https://github.com/sagemathinc/cocalc/assets/1609739/dd4ed057-ad11-44ae-9027-a0742d77bca6)

* `handleMessage` sends good looking json on `handleMessage` (`src/packages/project/named-servers/index.ts`)

![image](https://github.com/sagemathinc/cocalc/assets/1609739/185bafc0-6b24-43b1-9c9e-937d14b3836b)

* Emmm...  something wrong on processing it on hub side or something (I am very bad in debugging nodejs :( ). 
* I tried to bisect commits to find working («with VS Code») version, but it takes a lot of time, so I just report here as «reproduceable with cocalc-docker».

**Comment by williamstein** [comment]
For what it is worth I use vscode from cocalc docker nearly every day, so
maybe the problem is somehow related to something about your configuration?

-- William Stein

**Comment by belonesox** [comment]
May be.  :(

Main difference to explore — I use podman version of docker, because I use Fedora (on fedora linuxes docker-podman is preferrable default docker). 

If someone called it as written in documentation («just docker without sudo») it has «cocalc problem on running projects» because default docker podman used rootless containers with limitation of 65K subusers id (see `export function getUid(project_id: string)` on `src/packages/backend/misc.ts`) — on rootless containers we cannot use userids > 65K ).  But when called with sudo, all looks working except running «named servers» (VS Code, Jupyter Lab).

Same problem I see on regular installation on Fedora, remote or local, but cannot get what broken.

**Comment by belonesox** [comment]
Ubuntu 22.04 + docker.io  →  OK   :(

**Comment by belonesox** [comment]
I see the problem: 
* on Fedoras (even with podman-docker with cocalc), "localhost" resolves to "::1" (on Node builtin `http` module)
* on Ubuntus (or with moby-docker with cocalc), same "localhost" resolves to "127.0.0.1" 

* Nameserver Cocalc proxies map "/xxxx" to  "localhost:NNN" 
* Code-server listen only '127.0.0.1' and we got 

```
2023-10-26T19:37:52.885Z:   [33;1mcocalc:silly:proxy:handle-request [0m?folder=/mnt/work4/cocalc-projects/12c7c1f1-83c2-400c-b110-750eacb5f810 http proxy error event (ending proxy) {"errno":-111,"code":"ECONNREFUSED","syscall":"connect","address":"::1","port":38675} [33m+59s[0m
```

I try to figure out how to do right fix.

**Comment by belonesox** [comment]
because cocalc already have hardcoding on `project/named-servers/control.ts`

```js
…
let ip = INFO.location.host ?? "127.0.0.1";
if (ip == "localhost") {
    ip = "127.0.0.1";
 }
…  
```

I think, we should hardcode binding proxies for named servers at `projectControl` on `hub/proxy/target.ts`

**Comment by williamstein** [comment]
I agree with your fix. I personally don't use any computers or containers with ipv6, so haven't hit this problem.  However, some of our customers do, and I very much want cocalc (and cocalc-docker) to fully at least work in an ipv6 environment.   Thanks!

**Comment by belonesox** [comment]
Looks  OK, done here, thanks!

**Comment by ColossalFossil** [comment]
I found this closed issue when searching for podman issues related to cocalc-docker.  I experienced a similar issue, related to ipv4 / ipv6 binding.  I have not yet applied the code from this PR, but in addition to launching servers (VSCode, Jupyter, etc), my issue also affected Sage and any apps launched through an x11 worksheet.

I am hosting the cocalc-v2 aarch64 image in a podman container on Oracle Ampere compute.  Starting at the end, the default hosts file leads to inconvenient bindings.  Removing the default _localhost_  entry under _::1 _ resulted in functional bindings.  I do not know if there will be any additional side effects from this change, but so far, so good.

The default hosts file on oracle linux includes localhost twice:
127.0.0.1       **localhost** localhost.localdomain localhost4 localhost4.localdomain4
::1     **localhost** localhost.localdomain localhost6 localhost6.localdomain6

I spun up cocalc-v2-aarch64 in this environment through podman using this command (volume was optional/root was not):

_sudo podman run -h www.someSampleFakeDomain.org -e --rm --name=cocalc -d -v cocalc_state:/projects:Z -p 443:443 sagemathinc/cocalc-v2-aarch64_

image used
docker.io/sagemathinc/cocalc-v2-aarch64  latest      f9dbf5f2ccde  3 weeks ago   21.5 GB

A copy of the system hosts file forms the basis of the podman container hosts file:

127.0.0.1       **localhost** localhost.localdomain localhost4 localhost4.localdomain4
::1    **localhost** localhost.localdomain localhost6 localhost6.localdomain6
10.0.0.225      instanceSomeNumbers.subnetMoreNums.vcnMoreNums.oraclevcn.com instanceSomeNumbers
10.88.0.1       host.containers.internal
10.88.0.3       www.someSampleFakeDomain.org cocalc

This resulted in some bindings looking like this:

~$ netstat -l
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp        0      0 0.0.0.0:https           0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:ssh             0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:http            0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:15300           0.0.0.0:*               LISTEN     
tcp6       0      0 localhost:35515         [::]:*                  LISTEN     
tcp6       0      0 [::]:ssh                [::]:*                  LISTEN     
tcp6       0      0 localhost:33407         [::]:*                  LISTEN     
tcp6       0      0 localhost:44057         [::]:*                  LISTEN

The site started up fine, and the terminal was accessible.  Sage socket retrieval just timed out; the x11 window had rapid-fire port failures while trying to connect before launching any apps.

Removing the ipv6 loopback for _localhost_ gave podman a new starting point for its hosts file:
127.0.0.1       **localhost** localhost.localdomain localhost4 localhost4.localdomain4
::1     localhost.localdomain localhost6 localhost6.localdomain6

This resulted in working bindings, and a fully-functional CoCalc-docker deployment:
~$ netstat -l
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp        0      0 0.0.0.0:https           0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:http            0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:ssh             0.0.0.0:*               LISTEN     
tcp        0      0 localhost:46415         0.0.0.0:*               LISTEN     
tcp        0      0 localhost:34915         0.0.0.0:*               LISTEN     
tcp        0      0 localhost:35613         0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:16076           0.0.0.0:*               LISTEN     
tcp6       0      0 [::]:ssh                [::]:*                  LISTEN 

I tried disabling ipv6 and other more complicated solutions before finding the hosts issue.  Incidentally, hot-editing the container host file to remove the ::1 localhost entry allowed connectivity to the services, _after_ logging in to the main site with the original hosts file, which was entertaining.  Changing it back and forth toggled which parts of CoCalc would function.

This is a really cool project and I can't wait to play with it more.  Sorry for abusing the source-control-issue-system.
